File: modules/kidi-live-effects/android/src/main/java/com/kidiplus/liveeffects/KidiLiveEffectsModule.kt
package com.kidiplus.liveeffects

import expo.modules.kotlin.exception.Exceptions
import expo.modules.kotlin.functions.Coroutine
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

class KidiLiveEffectsModule : Module() {
    private val sessionLock = Any()
    private var session: KidiLiveEffectsSession? = null

    private fun requireSession(): KidiLiveEffectsSession = synchronized(sessionLock) {
        session ?: buildSession().also { session = it }
    }

    private fun buildSession(): KidiLiveEffectsSession {
        val reactContext = appContext.reactContext ?: throw Exceptions.ReactContextLost()
        return KidiLiveEffectsSession(
            context = reactContext,
            activityProvider = { appContext.currentActivity },
        ).also {
            it.listener = object : KidiLiveEffectsSession.Listener {
                override fun onUnavailable() {
                    sendEvent("unavailable", emptyMap<String, Any?>())
                }

                override fun onFirstFrame() {
                    sendEvent("firstFrame", emptyMap<String, Any?>())
                }
            }
            KidiLiveEffectsSessionHolder.session = it
            KidiLiveEffectsSessionHolder.previewHost?.let { host -> it.attachPreview(host) }
        }
    }

    override fun definition() = ModuleDefinition {
        Name("KidiLiveEffects")

        Events("unavailable", "firstFrame")

        OnActivityEntersBackground {
            session?.onHostPause()
        }

        OnActivityEntersForeground {
            session?.onHostResume()
        }

        OnDestroy {
            session?.onHostDestroy()
            if (KidiLiveEffectsSessionHolder.session === session) {
                KidiLiveEffectsSessionHolder.session = null
            }
            session = null
        }

        AsyncFunction("warmup") Coroutine {
            mapOf("supported" to requireSession().warmup())
        }

        AsyncFunction("start") Coroutine { config: Map<String, Any?> ->
            mapOf("started" to requireSession().start(config))
        }

        AsyncFunction("setConfig") Coroutine { config: Map<String, Any?> ->
            requireSession().setConfig(config)
            mapOf("updated" to true)
        }

        AsyncFunction("stop") Coroutine {
            session?.stop()
            mapOf("stopped" to true)
        }

        View(KidiLiveEffectsPreviewView::class) {}
    }
}

File: modules/kidi-live-effects/ios/KidiLiveEffectsModule.swift
import ExpoModulesCore
import UIKit

public class KidiLiveEffectsModule: Module {
  public func definition() -> ModuleDefinition {
    Name("KidiLiveEffects")

    Events("unavailable", "firstFrame")

    OnCreate {
      KidiLiveEffectsSession.shared.onUnavailable = { [weak self] in
        self?.sendEvent("unavailable", [:])
      }
      KidiLiveEffectsSession.shared.onFirstFrame = { [weak self] in
        self?.sendEvent("firstFrame", [:])
      }
    }

    OnDestroy {
      KidiLiveEffectsSession.shared.onUnavailable = nil
      KidiLiveEffectsSession.shared.onFirstFrame = nil
    }

    View(KidiLiveEffectsPreviewView.self)

    AsyncFunction("warmup") { (promise: Promise) in
      KidiLiveEffectsSession.shared.warmup { ok in
        promise.resolve(["supported": ok])
      }
    }

    AsyncFunction("start") { (config: [String: Any], promise: Promise) in
      KidiLiveEffectsSession.shared.start(config: config) { started in
        promise.resolve(["started": started])
      }
    }

    AsyncFunction("setConfig") { (config: [String: Any], promise: Promise) in
      KidiLiveEffectsSession.shared.setConfig(config) {
        promise.resolve(["updated": true])
      }
    }

    AsyncFunction("stop") { (promise: Promise) in
      KidiLiveEffectsSession.shared.stop {
        promise.resolve(["stopped": true])
      }
    }
  }
}
